content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -579,41 +657,919 @@ sending values over the wire, the code should query the size of the
+record_representation_clause completely specifies the representation of 
+the type and will be obeyed exactly by the implementation. 
+
+Components not otherwise given an explicit placement are given positions 
+chosen by the compiler. We don't say that they "follow" those explicitly 
+positioned because there's no requirement that the explicit positions 
+start at offset 0, although it would be unusual not to start there. 
+
+Placements must not make components overlap, except for components of 
+variant parts, a topic covered elsewhere. You can also specify the 
+placement of implementation-defined components, as long as you have a 
+name to refer to them. (In addition to the components listed in the 
+source code, the implementation can add components to help implement 
+what you wrote explicitly.) Such names are always attribute 
+references but the specific attributes, if any, are 
+implementation-defined. It would be a mistake for the compiler to define 
+such implicit components without giving you a way to refer to them. 
+Otherwise they might go exactly where you want some other component to 
+be placed, or overlap that place. 
+
+The positions (offsets) and the bit numbers must be static, informally 
+meaning that they are known at compile-time. They don't have to be 
+numeric literals, though. Numeric constants would work, but literals are 
+the most common by far. 
+
+Note that the language does not limit support for component clauses to 
+specific component types. They need not be one of the integer types, in 
+particular. For example, a position can be given for components that are 
+themselves record types, or array types. Even task types are allowed as 
+far as the language goes, although the implementation might require a 
+specific representation, such as the component taking no bits whatsoever 
+(:ada:`0 .. -1`). There are restrictions that keep things sane, for 
+example rules about how a component name can be used within the overall 
+record layout construct, but not restrictions on the types allowed for 
+individual components. For example, here is a record layout containing a 
+String component, arbitrarily set to contain 11 characters: 
+
+.. code-block:: ada
+
+   type R is record
+      S : String (1 .. 11);
+      B : Boolean;
+   end record;
+
+   for R use record
+      S at 0  range 0 .. 87;
+      B at 11 range 0 .. 7;
+   end record;
+   
+Component S is to be the first component in memory in this example, 
+hence the position offset is 0, for the first byte of S. Next, S is 11 
+characters long, or 88 bits, so the bit range is 0 .. 87. That's 11 
+bytes of course, so S occupies storage elements 0 .. 10. Therefore, the 
+next component position must be at least 11, unless there is to be a 
+gap, in which case it would be greater than 11. We'll place B 
+immediately after the last character of S, so B is at storage element 
+offset 11 and occupying all that one byte's bits. 
```

Comment:
```suggestion
Component :ada:`S` is to be the first component in memory in this example, 
hence the position offset is 0, for the first byte of :ada:`S`. Next, :ada:`S` is 11 
characters long, or 88 bits, so the bit range is 0 .. 87. That's 11 
bytes of course, so :ada:`S` occupies storage elements 0 .. 10. Therefore, the 
next component position must be at least 11, unless there is to be a 
gap, in which case it would be greater than 11. We'll place :ada:`B` 
immediately after the last character of :ada:`S`, so :ada:`B` is at storage element 
offset 11 and occupying all that one byte's bits. 
```